WordPress/src/main/java/org/wordpress/android/ui/posts/mediauploadcompletionprocessors/CoverBlockProcessor.java

```diff
@@ -0,0 +1,78 @@
+package org.wordpress.android.ui.posts.mediauploadcompletionprocessors;
+
+import com.google.gson.JsonObject;
+
+import org.jsoup.nodes.Document;
+import org.jsoup.nodes.Element;
+import org.wordpress.android.util.helpers.MediaFile;
+
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+public class CoverBlockProcessor extends BlockProcessor {
+    /**
+     * Template pattern used to match and splice cover inner blocks
+     */
+    private static final Pattern PATTERN_COVER_INNER = Pattern.compile(new StringBuilder()
+                    .append("(^.*?<div class=\"wp-block-cover__inner-container\">\\s*)")
+                    .append("(.*)") // inner block contents
+                    .append("(\\s*</div>\\s*</div>\\s*<!-- /wp:cover -->.*)").toString(), Pattern.DOTALL);
+
+    /**
+     * Pattern to match background-image url in cover block html content
+     */
+    private static final Pattern PATTERN_BACKGROUND_IMAGE_URL = Pattern.compile("background-image:url\\([^\\)]+\\)");
+
+    private final MediaUploadCompletionProcessor mMediaUploadCompletionProcessor;
+
+    public CoverBlockProcessor(String localId, MediaFile mediaFile,
+                               MediaUploadCompletionProcessor mediaUploadCompletionProcessor) {
+        super(localId, mediaFile);
+        mMediaUploadCompletionProcessor = mediaUploadCompletionProcessor;
+    }
+
+    @Override String processInnerBlock(String block) {
+        Matcher innerMatcher = PATTERN_COVER_INNER.matcher(block);
+        boolean innerCapturesFound = innerMatcher.find();
+
+        // process inner contents recursively
+        if (innerCapturesFound) {
+            String innerProcessed = mMediaUploadCompletionProcessor.processContent(innerMatcher.group(2)); //
+            return new StringBuilder()
+                    .append(innerMatcher.group(1))
+                    .append(innerProcessed)
+                    .append(innerMatcher.group(3))
+                    .toString();
+        }
+
+        return block;
+    }
+
+    @Override boolean processBlockJsonAttributes(JsonObject jsonAttributes) {
+        if (jsonAttributes.get("id").getAsInt() == Integer.parseInt(mLocalId, 10)) {
+            jsonAttributes.addProperty("id", Integer.parseInt(mRemoteId, 10));
+            jsonAttributes.addProperty("url", mRemoteUrl);
+            return true;
+        }
+
+        return false;
+    }
+
+    @Override boolean processBlockContentDocument(Document document) {
+        // select cover block div
+        Element targetDiv = document.select(".wp-block-cover").first();
+
+        // if a match is found, proceed with replacement
+        if (targetDiv != null) {
+            // replace background-image url in style attribute
+            String style = PATTERN_BACKGROUND_IMAGE_URL.matcher(targetDiv.attr("style"))
+                    .replaceFirst(String.format("background-image:url(%1$s)", mRemoteUrl));
+            targetDiv.attr("style", style);
```

Comment:
hm…I'm seeing the cover block actually store the background imagine inside the JSON attributes. this will deifnitely update the style, but won't the newly-uploaded image get wiped out after someone loads the post in the editor again and there's a mismatch between what the attributes indicate and what HTML has been generated?

we probably need to update both places.

Reply:
> we probably need to update both places

The Json attributes are updated in an [earlier step](https://github.com/wordpress-mobile/WordPress-Android/pull/11541/files/e6c78af481ba7d6d361d9ff526933494bac0ee32#diff-5cc438e830db3e4f6b106dca92c585c8R91) via the abstract method `processBlockJsonAttributes`. For Cover block, the `id` is used to check for a match first, and the `url` [is mutated](https://github.com/wordpress-mobile/WordPress-Android/pull/11541/files/e6c78af481ba7d6d361d9ff526933494bac0ee32#diff-8c06e672b915270d7915261d8e7e367cR54) a match with the local id is found. The replacement within the `style` attribute only occurs subsequently, if that condition is met.